services/stats/memstats/stats.go

```diff
@@ -0,0 +1,181 @@
+package memstats
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/services/stats"
+)
+
+var _ stats.Stats = (*Store)(nil)
+
+var _ stats.Measurement = (*Measurement)(nil)
+
+type Store struct {
+	once  sync.Once
+	mu    sync.Mutex
+	byKey map[string]*Measurement
+
+	Now func() time.Time
+}
+
+type Measurement struct {
+	mu        sync.Mutex
+	startTime time.Time
+	now       func() time.Time
+
+	Tags stats.Tags
+	Name string
+	Type string
+
+	LastValue    float64
+	Values       []float64
+	LastDuration time.Duration
+	Durations    []time.Duration
+}
+
+// Count implements stats.Measurement
+func (m *Measurement) Count(n int) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.Type != stats.CountType {
+		panic("operation Count not supported for measurement type:" + m.Type)
+	}
+
+	m.LastValue += float64(n)
+	m.Values = append(m.Values, m.LastValue)
+}
+
+// Increment implements stats.Measurement
+func (m *Measurement) Increment() {
+	if m.Type != stats.CountType {
+		panic("operation Increment not supported for measurement type:" + m.Type)
+	}
+
+	m.Count(1)
+}
+
+// Gauge implements stats.Measurement
+func (m *Measurement) Gauge(value interface{}) {
+	if m.Type != stats.GaugeType {
+		panic("operation Gauge not supported for measurement type:" + m.Type)
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.LastValue = value.(float64)
+	m.Values = append(m.Values, m.LastValue)
+}
+
+// Observe implements stats.Measurement
+func (m *Measurement) Observe(value float64) {
+	if m.Type != stats.HistogramType {
+		panic("operation Observe not supported for measurement type:" + m.Type)
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.LastValue = value
+	m.Values = append(m.Values, m.LastValue)
+}
+
+// Start implements stats.Measurement
+func (m *Measurement) Start() {
+	if m.Type != stats.TimerType {
+		panic("operation Start not supported for measurement type:" + m.Type)
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.startTime = m.now()
+}
+
+// End implements stats.Measurement
+func (m *Measurement) End() {
+	if m.Type != stats.TimerType {
+		panic("operation End not supported for measurement type:" + m.Type)
+	}
+
+	m.SendTiming(m.now().Sub(m.startTime))
+}
+
+// Since implements stats.Measurement
+func (m *Measurement) Since(start time.Time) {
+	if m.Type != stats.TimerType {
+		panic("operation Since not supported for measurement type:" + m.Type)
+	}
+
```

Comment:
This check is redundant.

Reply:
> 1. please check moq, let me know if it can help somehow

Overall I prefer moq for generating mocks. I still find it will add unnecessary complexity to the alerts. The reason is stats and measurement are very coupled, the problem appears when you have to deal with multiple metrics. Essentially the logic that is implemented here should be repeated one way or the other, in all the test cases. 

* Ensuring that the correct measurement (name, tags) is changed.
* Ignoring stats we are not interested in.

Thought I could have used the moq, to build the measurement implementation.

Given the frequency and complexity of testing stats, I would go ahead with this custom implementation. Maybe we should have the same for logging.


> 2. beware of reading `LastValue` directly in the tests. if some code is still running the write is protected by the library via a mutex but in the test you're reading it without passing through the mutex. it's usually fine in unit tests, less fine in integration tests where the code that you're testing might still be running while you make your assertions. that could cause a failure when running tests of that kind with `-race`
> 
> In any case just reading the last value in some cases feels a bit fragile so perhaps a race failure is OK as it would force us to review the test. Anyhow, this could be a good starting point, imo we can add more on it if needed down the road.

Good point, it slipped my mind, I will make sure it is thread-safe.